### PR TITLE
Implemented SmoothGrad

### LINKED ISSUE #56: Implement SmoothGrad
SmoothGrad is a very simple, but effective technique for computing saliency maps. Given an input x, it generates samples around x by adding Gaussian noise and then computing the saliency map on the sample. Then, it averages the saliency maps together.

### PR DESCRIPTION
**Pull request by ihowell**
Implemented SmoothGrad saliency maps.

Fixes #56

### CODE REVIEW

File: pyroclast/common/visualizable.py [hunk 1]
```diff
@@ -98,3 +98,25 @@ def sensitivity_map(self, x, softmax=False):
...
+
+    def smooth_grad(self, x, n=50, sigma=0.01):
+        """Computes the smooth grad saliency-map
+
+        SmoothGrad works by averaging the gradient of n input images
+        that have been perturbed with Gausssian noise. Gaussian noise
+        is parameterized by the standard deviation (sigma).
+
+        Args:
+           x (np.array): shape [batch_size, ...data_shape]
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: What is `x` though? I assume it should be some gradients, but it's never mentioned here

File: pyroclast/common/visualizable.py [hunk 2]
```diff
@@ -98,3 +98,25 @@ def sensitivity_map(self, x, softmax=False):
                 result = tf.nn.softmax(result)
             grads = tape.gradient(result, x)
         return grads
+
+    def smooth_grad(self, x, n=50, sigma=0.01):
+        """Computes the smooth grad saliency-map
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: Isn't "SmoothGrad" one word with specific capitalization? Further, if this implementation is based on a paper, you should cite it here.

File: pyroclast/common/visualizable.py [hunk 3]
```diff
@@ -98,3 +98,25 @@ def sensitivity_map(self, x, softmax=False):
...
+        is parameterized by the standard deviation (sigma).
+
+        Args:
+           x (np.array): shape [batch_size, ...data_shape]
+           n (int): The number of samples to collect
+           sigma (float): The standard deviation of the Gaussian noise
+
+        Returns:
+           sensitivity_map (np.array): shape [batch_size, ...data_shape]
+
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: Why the whitespace here?

File: pyroclast/common/visualizable.py [hunk 4]
```diff
@@ -98,3 +98,25 @@ def sensitivity_map(self, x, softmax=False):
...
+        """
+        gradients = np.array(
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: Why do this in numpy rather than TF?

File: pyroclast/common/visualizable_test.py [hunk 5]
```diff
@@ -61,6 +61,14 @@ def test_sensitivity_map(self):
...
-            input_sensitivity_maps = self.model.sensitivity_map(x, 0)
+            input_sensitivity_maps = self.model.sensitivity_map(x)
+            assert input_sensitivity_maps is not None
+            assert input_sensitivity_maps.shape[0:3] == x.shape[0:3]
+
+    def test_smooth_grad_map(self):
+        for batch in self.ds['train']:
+            x = tf.cast(batch['image'], tf.float32)
+            y = self.model(x)
+            input_sensitivity_maps = self.model.smooth_grad(x, n=3)
```
Comment (DrKwint) [inline pyroclast/common/visualizable_test.py]: If you're inputting the image, what does this calculate?

File: pyroclast/common/visualizable.py (hunk 1 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: x is a batch of input data. The method is constructed so that we can create the SmoothGrad maps for all inputs at once.

File: pyroclast/common/visualizable.py (hunk 4 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: True. I wanted to use the old saliency map method, but it would probably be more efficient to do the whole thing in TF.

File: pyroclast/common/visualizable_test.py (hunk 5 above)
Comment (ihowell) [reply pyroclast/common/visualizable_test.py]: We are inputting batches of images.

File: pyroclast/common/visualizable.py (hunk 1 above)
Comment (DrKwint) [reply pyroclast/common/visualizable.py]: I think that should be added to the docs before a merge. If you would.

File: pyroclast/common/visualizable.py (hunk 4 above)
Comment (DrKwint) [reply pyroclast/common/visualizable.py]: It should be a straightforward change to update that for the merge.

File: pyroclast/common/visualizable.py (hunk 1 above)
Comment (DrKwint) [reply pyroclast/common/visualizable.py]: If you're just taking in input data, then the function as written is simply calculating an empirical average of noised data centered at the input. There isn't actually a gradient calculation in there anywhere.

File: pyroclast/common/visualizable.py (hunk 4 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: Just competed this.

File: pyroclast/common/visualizable.py (hunk 1 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: I didn't even do the correct implementation this the first time.
